src/core/ensemble_scheduler.cc

```diff
@@ -0,0 +1,245 @@
+// schedule new steps if available and the last step will finish the ensemble
+// request.
+// So we don't have to maintian the context in scheduler as the shared_ptr
+// will destroy the context for us if there are no "in-flight" steps.
+class EnsembleContext {
+ public:
+  EnsembleContext(
+      InferenceServer* is, const ModelConfig& config,
+      const std::unordered_map<std::string, EnsembleTensor>& ensemble_graph,
+      const std::shared_ptr<ModelInferStats>& stats,
+      const std::shared_ptr<InferRequestProvider>& request_provider,
+      const std::shared_ptr<InferResponseProvider>& response_provider,
+      std::function<void(Status)> OnComplete);
+
+  // Initiate transition on 'context' state
+  static void Proceed(const std::shared_ptr<EnsembleContext>& context);
+
+ private:
+  using StepList = std::vector<std::shared_ptr<EnsembleStep>>;
+
+  // Perform transition on 'context' state given the information of
+  // 'completed_step'
+  static void Proceed(
+      const std::shared_ptr<EnsembleContext>& context,
+      const std::shared_ptr<EnsembleStep>& completed_step);
+
+  // Helper function that updates ensemble state given 'completed_step'
+  void UpdateEnsembleState(const std::shared_ptr<EnsembleStep>& completed_step);
+
+  // Helper function that returns a list of step that should be run
+  // under current state.
+  // 'ok' will be set to false if the ensemble context will not make any
+  // progress from current state. 'ok' will be set to true otherwise
+  StepList Next(bool& ok);
+
+  // Helper function that completes the response of the ensemble request
+  void FinishEnsemble();
+
+  // Helper function that prepares and calls the inference server's function
+  // to process infer requests specified in 'steps'
+  static void ScheduleSteps(
+      const std::shared_ptr<EnsembleContext>& context, const StepList& steps);
+
+  InferenceServer* is_;
+  // [TODO] we don't really need model config, we can construct data structure
+  //   for each step based on the model config in constructor
+  ModelConfig config_;
+  std::unordered_map<std::string, EnsembleTensor> ensemble_graph_;
+  std::shared_ptr<ModelInferStats> stats_;
+  std::shared_ptr<InferRequestProvider> request_provider_;
+  std::shared_ptr<InferResponseProvider> response_provider_;
+  std::function<void(Status)> OnComplete_;
+};
+
+EnsembleContext::EnsembleContext(
+    InferenceServer* is, const ModelConfig& config,
+    const std::unordered_map<std::string, EnsembleTensor>& ensemble_graph,
+    const std::shared_ptr<ModelInferStats>& stats,
+    const std::shared_ptr<InferRequestProvider>& request_provider,
+    const std::shared_ptr<InferResponseProvider>& response_provider,
+    std::function<void(Status)> OnComplete)
+    : is_(is), config_(config), ensemble_graph_(ensemble_graph), stats_(stats),
+      request_provider_(request_provider),
+      response_provider_(response_provider), OnComplete_(OnComplete)
+{
```

Comment:
At some point, perhaps here, you need to grab servable handles (backends) for all the models you will need in the ensemble. Otherwise they could be unloaded while the ensemble is executing.